add materialized view configuration for postgres (#4590)
fishtown-analytics/docs.getdbt.com: website/docs/reference/resource-configs/postgres-configs.md

    [+](/reference/resource-configs/plus-prefix)on_configuration_change: apply | continue | fail
    [+](/reference/resource-configs/plus-prefix)[indexes](#indexes):
      - columns: [<column-name>]
        unique: true | false
        type: hash | btree
```

</File>

</TabItem>

<TabItem value="property-yaml">

<File name='models/properties.yml'>

```yaml
version: 2

models:
  - name: [<model-name>]
    config:
      [materialized](/reference/resource-configs/materialized): materialized_view
      on_configuration_change: apply | continue | fail
      [indexes](#indexes):
        - columns: [<column-name>]
          unique: true | false
          type: hash | btree
```

</File>

</TabItem>

<TabItem value="config">

<File name='models/<model_name>.sql'>

```jinja
{{ config(
    [materialized](/reference/resource-configs/materialized)="materialized_view",
    on_configuration_change="apply" | "continue" | "fail",
    [indexes](#indexes)=[
        {
            "columns": ["<column-name>"],
            "unique": true | false,
            "type": "hash" | "btree",
        }
    ]
) }}
```

</File>

</TabItem>

</Tabs>

> The `indexes` parameter corresponds to that of a table, as linked above.

Reviewer: ```suggestion
The `indexes` parameter corresponds to that of a table, as linked above.
```
```suggestion
The [`indexes`](#indexes) parameter corresponds to that of a table, as explained above.
```

> It's worth noting that, unlike with tables, dbt will monitor this parameter for changes and apply the changes without dropping the materialized view.

Reviewer: ```suggestion
It's worth noting that, unlike with tables, dbt monitors this parameter for changes and applies the changes without dropping the materialized view.
```

> It's worth noting that, unlike with tables, dbt monitors this parameter for changes and applies the changes without dropping the materialized view.

Reviewer: ```suggestion
It's worth noting that, unlike tables, dbt monitors this parameter for changes and applies the changes without dropping the materialized view.
```

> This happens via a `DROP/CREATE` of the indexes, which could be thought of as a `ALTER` of the materialized view.

Reviewer: ```suggestion
This happens via a `DROP/CREATE` of the indexes, which can be thought of as an `ALTER` of the materialized view.
```

## Materialized views

The Postgres adapter supports [materialized views](https://www.postgresql.org/docs/current/rules-materializedviews.html)
with the following configuration parameters:

| Parameter                 | Type       | Required | Default   | Change Monitoring Support | Reference           |
|---------------------------|------------|----------|-----------|---------------------------|---------------------|

> | `on_configuration_change` | STRING     | NO       | `'apply'` | N/A                       |                     |

Reviewer 2: Similarly to the link to "Indexes", do you want to link to the docs on `on_configuration_change ` that are featured [here](https://docs.getdbt.com/docs/build/materializations#materialized-view)?

Author: Honestly, I want to break that out as its own section, and then link to that both here and in materialized views. It probably makes sense to add the link here for now though.

Author: It turns out that I didn't create that as a link-able section. Instead of updating that just to link it, I'd like to create a separate PR after these four are in and update all of these non-links to link to the new section.

> | `indexes`                 | LIST[DICT] | NO       |           | ALTER                     | [Indexes](#indexes) |

Reviewer 2: The table can be one column thinner if doing a hyperlink like [`indexes`](#indexes) in the first column instead.

Author: Fair. Let me play with it and see how it looks. When I started the docs updates I didn't have a link for everything, so it felt like an after thought. but at this point, I think we do have one, so it's probably better to have it up front.

Author: Done

#### Sample model file:

<File name='postgres_materialized_view.sql'>

```sql
{{ config(
    materialized='materialized_view',
    on_configuration_change='{ apply | continue | fail }'
    indexes=[
        {
            'columns': ['<column_name>', ...],

>             'unique': <bool>},

Reviewer 2: Looks like there is an `}` here when it is intended to be lower down.
```suggestion
            'unique': <bool>,
```

>             'type': '{ HASH | B-TREE | GIST | SP-GIST | GIN | BRIN }',

Reviewer 2: Did we test if all six of these index types actually work?

Curious if they are supposed to be 
 [`btree`, `hash`, `gist`, `spgist`, `gin`, and `brin`](https://github.com/dbt-labs/dbt-core/blob/5488dfb9926c3f81e6a80bd0d8b01fe7f0ac7a60/plugins/postgres/dbt/adapters/postgres/relation_configs/index.py#L19-L22), by any chance.
 
 Based on https://github.com/dbt-labs/dbt-core/issues/9185, also wondering if  `gist`, `spgist`, `gin`, and `brin` work as-is, or if they need an operator class ("opclass") specified during index creation in order to work?

Author: Oh, fair point. I just assumed from the original index docs that any type was accepted; it doesn't specify nor limit the list, instead it points directly to the Postgres docs and says "a valid type from ...". I can limit this list if we only support a subset.

Author: I reduced this to just `hash` and `btree`, which I know work from tests I wrote for materialized views.

>     ]

Reviewer 2: Guessing there's intended to be a `}` here right before `]`.
```suggestion
    }]
```

) }}

select * from {{ ref('my_base_table') }}
```

</File>

The `indexes` parameter corresponds to that of a table, as linked above.
It's worth noting that, unlike with tables, dbt will monitor this parameter for changes and apply the changes without dropping the materialized view.
This happens via a `DROP/CREATE` of the indexes, which could be thought of as a `ALTER` of the materialized view.

Find more information about materialized view parameters in the Postgres docs:
- [CREATE MATERIALIZED VIEW](https://www.postgresql.org/docs/current/sql-creatematerializedview.html)

> <VersionBlock lastVersion="1.6">

Reviewer 2: Did you want to do this the same as [#4582](https://github.com/dbt-labs/docs.getdbt.com/pull/4582/files) and [#4589](https://github.com/dbt-labs/docs.getdbt.com/pull/4589/files)? Or do you want it to be open-ended for all versions up to 1.6?

```suggestion
<VersionBlock firstVersion="1.6" lastVersion="1.6">
```

Author: I'm actually not sure how nested version blocks work. If they stack, I think we should only specify `lastVersion` since `firstVersion` is already specified in the outer block. Clearly I went back and forth on this since I wound up submitting both versions.

Author: After thinking about this, I would rather exclude `firstVersion`. Keeping just `lastVersion` aligns with the idea that this limitation was removed in 1.7. Adding `firstVersion` seems to suggest this limitation was introduced in 1.6, which is accurate, but misleading.
